src/coreclr/tools/Common/TypeSystem/Common/AsyncMethodVariant.cs

```diff
@@ -0,0 +1,88 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using Internal.TypeSystem.Ecma;
+
+namespace Internal.TypeSystem
+{
+    /// <summary>
+    /// Either the AsyncMethodImplVariant or AsyncMethodThunkVariant of a method marked .IsAsync.
+    /// </summary>
+    public partial class AsyncMethodVariant : MethodDelegator
+    {
+        private MethodSignature _asyncSignature;
+
+        public AsyncMethodVariant(EcmaMethod wrappedMethod)
+            : base(wrappedMethod)
+        {
+            Debug.Assert(wrappedMethod.Signature.ReturnsTaskOrValueTask());
+        }
+
+        public EcmaMethod Target => (EcmaMethod)_wrappedMethod;
+
+        public override MethodSignature Signature
+        {
+            get
+            {
+                return _asyncSignature ?? InitializeSignature();
+            }
+        }
+
+        private MethodSignature InitializeSignature()
+        {
+            var signature = _wrappedMethod.Signature;
+            Debug.Assert(!signature.IsAsyncCallConv);
+            Debug.Assert(signature.ReturnsTaskOrValueTask());
+            TypeDesc md = signature.ReturnType;
+            MethodSignatureBuilder builder = new MethodSignatureBuilder(signature);
+            builder.ReturnType = md.HasInstantiation ? md.Instantiation[0] : this.Context.GetWellKnownType(WellKnownType.Void);
+            builder.Flags = signature.Flags | MethodSignatureFlags.AsyncCallingConvention;
+            return (_asyncSignature = builder.ToSignature());
+        }
+
+        public override MethodDesc GetCanonMethodTarget(CanonicalFormKind kind)
+        {
+            return this;
+        }
+
+        public override MethodDesc GetMethodDefinition()
+        {
+            return this;
+        }
+
+        public override MethodDesc GetTypicalMethodDefinition()
+        {
+            return this;
```

Comment:
Lost a comment and assert from my PR version.

```suggestion
            // We should not be calling GetCanonMethodTarget on generic definitions of anything
            // and this MethodDesc is a generic definition.
            Debug.Assert(!HasInstantiation && !OwningType.HasInstantiation);
            return this;
```